OpenKeychain/src/main/java/org/sufficientlysecure/keychain/keyimport/WebKeyDirectoryClient.java

```diff
@@ -0,0 +1,203 @@
+                        "response:" + response);
+            }
+
+        } catch (IOException e) {
+            Timber.e(e, "IOException at Web Key Directory key download");
+            throw new QueryFailedException("Cannot connect to Web Key Directory. "
+                    + "Check your Internet connection!"
+                    + (proxy.getProxy() == Proxy.NO_PROXY ? "" : " Using proxy " + proxy.getProxy()));
+        }
+    }
+
+    @Override
+    public void add(String armoredKey, ParcelableProxy proxy) throws AddKeyException {
+        throw new UnsupportedOperationException("Uploading keys to Web Key Directory not supported");
+    }
+
+    /**
+     * Web Key Directory returns the entire key even during our searching phase.
+     *
+     * @throws UnsupportedOperationException if the key could not be parsed
+     */
+    @NonNull
+    private static ImportKeysListEntry getEntry(String name, byte[] data)
+            throws UnsupportedOperationException, IOException, PgpGeneralException {
+        ImportKeysListEntry entry = new ImportKeysListEntry();
+        entry.setSecretKey(false); // keys imported from Web Key Directory are public
+
+        UncachedKeyRing ring = UncachedKeyRing.decodeFromData(data);
+        UncachedPublicKey key = ring.getPublicKey();
+
+        entry.setUserIds(key.getUnorderedUserIds());
+        entry.setPrimaryUserId(key.getPrimaryUserIdWithFallback());
+
+        entry.setKeyId(key.getKeyId());
+        entry.setParcelableKeyRing(ParcelableKeyRing.createFromEncodedBytes(data));
+        entry.setFingerprint(key.getFingerprint());
+
+        try {
+            entry.setBitStrength(key.getBitStrength());
+            final int algorithm = key.getAlgorithm();
+            entry.setAlgorithm(KeyFormattingUtils.getAlgorithmInfo(algorithm, key.getBitStrength(),
+                    key.getCurveOid()));
+        } catch (NumberFormatException | NullPointerException e) {
+            Timber.e(e, "Conversion for bit size, algorithm, or creation date failed.");
+            // can't use this key
+            throw new UnsupportedOperationException(
+                    "Conversion for bit size, algorithm, or creation date failed.");
+        }
+
+        return entry;
+    }
+
+    private static byte[] toSHA1(byte[] input) {
+        try {
+            return MessageDigest.getInstance("SHA-1").digest(input);
+        }
+        catch (NoSuchAlgorithmException e) {
+            throw new AssertionError("SHA-1 should always be available");
+        }
+    }
+
+    private static final char[] ALPHABET = "ybndrfg8ejkmcpqxot1uwisza345h769".toCharArray();
+    private static final int SHIFT = Integer.numberOfTrailingZeros(ALPHABET.length);
+    private static final int MASK = ALPHABET.length - 1;
+
+    private static String encodeZBase32(byte[] data) {
+        if (data.length == 0) {
+            return "";
+        }
+
+        StringBuilder result = new StringBuilder();
```

Comment:
Maybe you can safetly initialize the size data.length() + SHIFT to avoid realloc?

Reply:
Honestly I'd keep it as it is. The code for ZBase32 is already quite opaque and this function is not inside any tight loop (it's called only once per search) so that the no-reallocating buffer version wouldn't show any observable benefits.